### PR TITLE
Enable CSI topology feature for K8S 1.22 and above

### PR DESCRIPTION
**Pull request by blc1996**
Signed-off-by: Lucheng Bao <luchengb@vmware.com>

### What this PR does / why we need it

CSI 2.4 introduced a ned label `topology-categories` that is not available in previous release, and it also can't be used in upgrade case when CSI enbaled beta multi-AZ feature before.

Only enable this `topology-categories` CSI feature in fresh install of K8S 1.22 and above. (K8S below 1.22 doesn't use CSI 2.4)

### Which issue(s) this PR fixes

Fixes # CSI 2.4 multi-AZ feature upgrade issue

### Describe testing done for PR

- [x] Manual testing
- [ ] Downstream pipeline

### Release note

```release-note
CSI package fixes the upgrade issue for multi-AZ topology feature.
```

### PR Checklist

- [x] Squash the commits into one or a small number of logical commits
      
- [x] Use good commit [messages](https://github.com/vmware-tanzu/tanzu-framework/blob/main/CONTRIBUTING.md)
- [x] Ensure PR contains terms all contributors can understand and links all contributors can access

### CODE REVIEW

File: pkg/v1/providers/ytt/02_addons/csi/csi_addon_data.lib.yaml [hunk 1]
```diff
@@ -1,6 +1,26 @@
 #@ load("@ytt:data", "data")
+#@ load("@ytt:regexp", "regexp")
+#@ load("@ytt:assert", "assert")
 #@ load("/lib/helpers.star", "get_bom_data_for_tkr_name", "get_image_repo_for_component", "get_no_proxy")
 
+
+#! Set topologyFeatureFlag to be true for K8S 1.22 and newer
+#@ def enableTopologyFeature():
+#@ if regexp.match("^v(0|[1-9]\\d*)\\.(0|[1-9]\\d*)\\.(0|[1-9]\\d*)(?:-((?:0|[1-9]\\d*|\\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\\.(?:0|[1-9]\\d*|\\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\\+([0-9a-zA-Z-]+(?:\\.[0-9a-zA-Z-]+)*))?$", data.values.KUBERNETES_VERSION) != True:
```
Comment (danniel1205) [inline pkg/v1/providers/ytt/02_addons/csi/csi_addon_data.lib.yaml:9]: Suggestion: Maybe let's use TKR version instead ? Although TKR version is usually aligned with K8S version, semantically whether passing in `true` of this flag is based on CSI version, and CSI version is decided by TKR. 

Using K8S version for comparison works, but people might wonder if this flag is to fix some compatibility issue between CSI and K8S, however they are actually unrelated.

File: pkg/v1/providers/ytt/02_addons/csi/csi_addon_data.lib.yaml [hunk 2]
```diff
@@ -56,6 +76,9 @@ vsphereCSI:
   http_proxy: #@ data.values.TKG_HTTP_PROXY
   https_proxy: #@ data.values.TKG_HTTPS_PROXY
   no_proxy: #@ get_no_proxy()
+  #@ if enableTopologyFeature():
+  useTopologyFeature: True
```
Comment (danniel1205) [inline pkg/v1/providers/ytt/02_addons/csi/csi_addon_data.lib.yaml:80]: Suggestion: will `useTopologyFeature : #@ enableTopologyFeature()` be a good idea ?

File: pkg/v1/providers/ytt/02_addons/csi/csi_addon_data.lib.yaml (hunk 1 above)
Comment (blc1996) [reply pkg/v1/providers/ytt/02_addons/csi/csi_addon_data.lib.yaml:9]: It seems that BOM file doesn't have a valid TKR version available.

e.g.
```
apiVersion: run.tanzu.vmware.com/v1alpha2
default:
  k8sVersion: v1.21.5+vmware.1-tkg.1-20211029
release:
  version: v1.5.0-20211029
```
We can only infer the TKR major, minor and patch version by K8S version.

File: pkg/v1/providers/ytt/02_addons/csi/csi_addon_data.lib.yaml (hunk 2 above)
Comment (blc1996) [reply pkg/v1/providers/ytt/02_addons/csi/csi_addon_data.lib.yaml:80]: Right, will change this line.

File: pkg/v1/providers/ytt/02_addons/csi/csi_addon_data.lib.yaml (hunk 1 above)
Comment (blc1996) [reply pkg/v1/providers/ytt/02_addons/csi/csi_addon_data.lib.yaml]: I was looking at TKG BOM, should look at TKR BOM instead. Changed the reference to TKR version.

File: pkg/v1/providers/ytt/02_addons/csi/csi_addon_data.lib.yaml [hunk 3]
```diff
@@ -1,6 +1,27 @@
 #@ load("@ytt:data", "data")
+#@ load("@ytt:regexp", "regexp")
+#@ load("@ytt:assert", "assert")
 #@ load("/lib/helpers.star", "get_bom_data_for_tkr_name", "get_image_repo_for_component", "get_no_proxy")
 
+
+#! Set topologyFeatureFlag to be true for TKR 1.22 and newer
+#@ def enableTopologyFeature():
+#@ bomData = get_bom_data_for_tkr_name()
+#@ if regexp.match("^v(0|[1-9]\\d*)\\.(0|[1-9]\\d*)\\.(0|[1-9]\\d*)(?:-((?:0|[1-9]\\d*|\\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\\.(?:0|[1-9]\\d*|\\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\\+([0-9a-zA-Z-]+(?:\\.[0-9a-zA-Z-]+)*))?$", bomData.release.version) != True:
```
Comment (vijaykatam) [inline pkg/v1/providers/ytt/02_addons/csi/csi_addon_data.lib.yaml:10]: (blocking comment): If we really need to use this regex here we'll need to add unit tests for templating that cover it along with documentation. Alternatively try to use [version](https://carvel.dev/ytt/docs/latest/lang-ref-ytt-version/) module

Comment (blc1996) [reply pkg/v1/providers/ytt/02_addons/csi/csi_addon_data.lib.yaml]: Removed regex mathing. Did some best effort validation for TKR versions.

**Review by vijaykatam — APPROVED** [review]
Change LGTM.

**Pull request merged**